engine/gamesys/src/gamesys/scripts/script_sys_gamesys.cpp

```diff
@@ -0,0 +1,476 @@
+    {
+        {"load_buffer",       Sys_LoadBuffer},
+        {"load_buffer_async", Sys_LoadBufferAsync},
+        {0, 0}
+    };
+
+    /*# an asyncronous request has finished successfully
+     * @name sys.REQUEST_STATUS_FINISHED
+     * @variable
+     */
+
+    /*# an asyncronous request is unable to read the resource
+     * @name sys.REQUEST_STATUS_ERROR_IO_ERROR
+     * @variable
+     */
+
+    /*# an asyncronous request is unable to locate the resource
+     * @name sys.REQUEST_STATUS_ERROR_NOT_FOUND
+     * @variable
+     */
+
+    void ScriptSysGameSysRegister(const ScriptLibContext& context)
+    {
+        lua_State* L = context.m_LuaState;
+        int top = lua_gettop(L);
+        (void)top;
+
+        luaL_register(L, "sys", ScriptImage_methods);
+
+    #define SETCONSTANT(name, val) \
+        lua_pushnumber(L, (lua_Number) val); \
+        lua_setfield(L, -2, #name);\
+
+        SETCONSTANT(REQUEST_STATUS_FINISHED,        REQUEST_STATUS_FINISHED);
+        SETCONSTANT(REQUEST_STATUS_ERROR_IO_ERROR,  REQUEST_STATUS_ERROR_IO_ERROR);
+        SETCONSTANT(REQUEST_STATUS_ERROR_NOT_FOUND, REQUEST_STATUS_ERROR_NOT_FOUND);
+    #undef SETCONSTANT
+
+        lua_pop(L, 1);
+        assert(top == lua_gettop(L));
+
+        g_SysModule.m_Factory           = context.m_Factory;
+        g_SysModule.m_LoadQueue         = dmLoadQueue::CreateQueue(context.m_Factory);
+        g_SysModule.m_LoadRequestsMutex = dmMutex::New();
+    }
+
+    bool ScriptSysGameSysUpdate(const ScriptLibContext& context)
+    {
+        dmMutex::ScopedLock lk(g_SysModule.m_LoadRequestsMutex);
+
+        bool result = true;
+        uint32_t request_count = g_SysModule.m_LoadRequests.Capacity();
+        for (int i = 0; i < request_count; ++i)
+        {
+            LuaRequest* request = g_SysModule.m_LoadRequests.GetByIndex(i);
+
+            if (request)
+            {
+                // If we tried to dispatch the request, but the queue was full, we try again
+                if (request->m_Status == REQUEST_STATUS_INITALIZED)
+                {
+                    DispatchRequest(request);
+                }
+                else if (request->m_Status == REQUEST_STATUS_FINISHED        ||
+                         request->m_Status == REQUEST_STATUS_ERROR_NOT_FOUND ||
+                         request->m_Status == REQUEST_STATUS_ERROR_IO_ERROR)
+                {
+                    result &= HandleRequestCompleted(request);
+                }
+                else
+                {
+                    void* buf;
+                    uint32_t size;
+                    dmLoadQueue::LoadResult load_result;
+                    if (dmLoadQueue::EndLoad(g_SysModule.m_LoadQueue, request->m_Request, &buf, &size, &load_result) == dmLoadQueue::RESULT_OK)
+                    {
+                        HandleRequestCompleted(request);
+                    }
+                }
+            }
+        }
+
+        return result;
```

Comment:
The result is only checked in the unit tests?

Reply:
Yes, I refactored this to call a function from the test to check the latest result instead.